luap.1
.TH LUAP "1" "27 Oct 2024" "luap 0.9" "User Commands"
.SH NAME
luap \- Lua prompt with pretty-printing and autocompletion
.SH SYNOPSIS
.B luap
[
.BI \-e "\| STMT\^"
]
[
.BI \-l "\| NAME\^"
]
[
.B \-p
]
[
.B \-v
]
[
.B \-i
]
[
.B \-h
]
[
.B "<SCRIPT>"
] ...
.SH DESCRIPTION
.PP
.B luaprompt
is both an interactive Lua prompt that can be used instead
of the official interpreter, as well as a module that provides a Lua
command prompt that can be embedded in a host application.  As a
standalone interpreter it provides many conveniences that are missing
from the official Lua interpreter.  As an embedded prompt, it's meant
for applications that use Lua as a configuration or interface language
and can therefore benefit from an interactive prompt for debugging or
regular use.
.LP
.B luaprompt
features:
.RS 2
.IP \(bu 2
Readline-based input with history and completion: In particular all
keywords, global variables and table accesses (with string or
integer keys) can be completed in addition to Readline's standard
file completion.  Module names are also completed, for modules
installed in the standard directories, and completed modules can
optionally be loaded.
.IP \(bu
Persistent command history (retained across sessions), as well as
recording of command results for future reference.
.IP \(bu
Proper value pretty-printing for interactive use: When an expression
is entered at the prompt, all returned values are printed
(prepending with an equal sign is not required).  Values are printed
in a descriptive way that tries to be as readable as possible.  The
formatting tries to mimic Lua code (this is done to minimize
ambiguities and no guarantees are made that it is valid code).
Additionally, each value is stored in a table for future reference.
.IP \(bu
Color highlighting of error messages and variable printouts.
.RE
.LP
.B luap
tries to mimic the standard Lua interpreter as much as possible
(currently only the -E flag is not supported) while providing the
extra set of features described above.  All option arguments start
with a single hyphen, followed by a letter.  Multiple consecutive
options can be specified by their respective letters following a
single hyphen.  Anything that does not follow this format is
interpreted, either as a script file, which is to be executed (if it
is the first non-option argument), or as an argument that is to be
passed to the script (for subsequent non-option arguments).  A single
hyphen is interpreted as a request to execute the standard input.
If no options are passed,
.B luap
enters interactive mode, as if the
.I \-i
option had been specified.

A file with custom configuration, or startup actions can be placed
either at \fI~/.luaprc.lua\fP, or alternatively at
\fI~/.config/luaprc.lua\fP.  The first of these to be found is loaded
and executed at startup.

.SH OPTIONS
.PP
.B luap
accepts the following options:
.PP
.PD 0
.TP 10
.BI \-e "\| STMT\^"
Execute string
.IR STMT .
.TP
.BI \-l "\| NAME\^"
Require library
.IR NAME .
.TP
.B \-p
Force plain, uncolored output.
.TP
.B \-v
Print version information.
.TP
.B \-i
Enter interactive mode.
.TP
.B \-h
Show this help message and exit.
.TP
.B \-\-
Stop processing arguments.  The rest of the line is interpreted a
script file to execute, optionally followed by arguments that should
be passed to the script.
.PD
.PP
When
.B luap
is used with LuaJIT, the following additional arguments are also understood:
.PP
.PD 0
.TP 10
.BI \-j "\| CMD\^"
Perform LuaJIT control command.
.TP
.BI \-O "\| OPT\^"
Control LuaJIT optimizations.
.PD

.SH CONFIGURATION
.BR luaprompt "'s"
and subsequently
.BR luap "'s"
functionality, can be customized at build-time.  See the
documentation accompanying the source code for more details.

Additionally, you can customize
.BR luap "'s"
command line interface, as it is based on
.BR "GNU Readline" .
This includes the ability to change behavior, such as how completion is performed, define additional keybindings or redefine existing ones, define custom macros and other possibilities.  It is therefore highly recommended to consult
.BR readline (3)
for more information.

Note that
.B luaprompt
sets its application name, so that you can have custom Readline
configuration, that will be effective for
.B luaprompt
only (thus not affecting every other Readline-enabled application on
your system).  You can achieve that using
.BR $if " ... " $endif
conditional constructs within Readline's initialization file.  Again, see
.BR readline (3)
for more information.


.SH "SEE ALSO"
.BR lua (1),
.BR luac (1),
.BR readline (3)

.SH AUTHOR
Dimitris Papavasileiou (dpapavas@protonmail.ch), Boris Nagaev
.SH "REPORTING BUGS"
Please report bugs at the luaprompt home page at https://github.com/dpapavas/luaprompt.


ChangeLog
2012-04-07  Dimitris Papavasileiou  <dpapavas@gmail.com>

	* First public release.

2012-04-16  Dimitris Papavasileiou  <dpapavas@gmail.com>

	* Released version 0.1.2 with the following changes:
	  - Fixed a bug preventing multi-line input with Lua 5.1.
	  - Pretty-printing of userdata honours the __tostring
	    metamethod.

2012-06-30  Dimitris Papavasileiou  <dpapavas@gmail.com>

	* Released version 0.2 with the following changes:
	  - Fixed a couple of completion and pretty-printing glitches.
	  - Added support for completing file names.

2013-08-14  Dimitris Papavasileiou  <dpapavas@gmail.com>

	* Released version 0.3 with the following changes:
	  - Changed pretty-printing to always honour __tostring
	    metamethods.
	  - Modified the completion generator to return file name
	    completion only if nothing else matches.

2014-09-02  Dimitris Papavasileiou  <dpapavas@gmail.com>

	* Released version 0.3.1 with the following changes:
	  - Fixed numerous memory leaks as well as a minor off-by-one
	    bug in the file name completion generator.  Sources now
	    hold up to scrutiny by Valgrind.
	  - Removed the use of wordexp() to expand the path to the
	    history file.  Its use probably caused more problems than
	    it solved due to its unavailability on some platforms.

2014-10-06  Dimitris Papavasileiou  <dpapavas@gmail.com>

	* Released version 0.4 with the following improvements:

	  - The auto-completer now adds certain useful suffixes to
	    completed symbol names depending on their type.  An opening
	    parenthesis is added to function values and either a dot
	    or an opening bracket is added to table values.

	  - Symbol names can be excluded from consideration during
	    completion if they start with a configurable prefix.  This
	    allows, for example, variables starting with an underscore
	    to be treated as "hidden".

	  - Multi-line statement handling has been improved by treating
	    the whole statement as a single line, when browsing through
	    the	command history.

2014-10-30  Dimitris Papavasileiou  <dpapavas@gmail.com>

	* Released version 0.5 with the following improvements:

	  - The auto-completer now completes module names based on
	    package.preload, pacakge.path and package.cpath.  Module
	    names are either completed within strings for use with
	    require or as a global variable, as if the module was
	    already loaded.  In the latter case the tab completer also
	    loads the module once it's fully completed and exports it
	    as a global variable so that further completions apply to
	    the module's table.

	  - Completion can be disabled individually for certain kinds
	    of symbols via compile-time switches.

2015-06-09  Dimitris Papavasileiou  <dpapavas@gmail.com>

	* Released version 0.6 with the following improvements:

	- Switched to LuaRocks-based building/distribution.
	- Converted luaprompt to a proper Lua module.
	- Added saving of results into a table.
	- Added support for completion based on __index being a table.
	- Fixed filename completions to allow continuing completion into
	  subdirectories.
	- Various bugfixes.

2015-11-19  Dimitris Papavasileiou  <dpapavas@gmail.com>

	* Released version 0.7 with the following improvements:

	- Luap now sets rl_readline_name, to allow luaprompt-specific
	  readline configuration.
	- Added globalization of already loaded modules, which are not in
	the globals table.
	- Added a Makefile to allow building and installing without
	Luarocks.
	- Added a manpage (not installable via LuaRocks).
	- Various bugfixes.

2023-07-25  Dimitris Papavasileiou  <dpapavas@protonmail.ch>

	* Released version 0.8 with the following improvements:

	- Luap now loads and executes a file at startup, allowing custom
	  configuration and startup actions.
	- Added support for dynamic prompts.

2024-10-27  Dimitris Papavasileiou  <dpapavas@protonmail.ch>

	* Released version 0.9 with the following improvements:

	- Added ESCAPE_STRINGS configuration option, controlling whether
	  strings results returned by luap_describe or printed out by the
	  prompt, are escaped or left unchanged.


luap.lua
#!/usr/bin/env lua

-- Copyright (C) 2012-2023 Dimitris Papavasiliou, Boris Nagaev
--
-- Permission is hereby granted, free of charge, to any person
-- obtaining a copy of this software and associated documentation
-- files (the "Software"), to deal in the Software without
-- restriction, including without limitation the rights to use, copy,
-- modify, merge, publish, distribute, sublicense, and/or sell copies
-- of the Software, and to permit persons to whom the Software is
-- furnished to do so, subject to the following conditions:
--
-- The above copyright notice and this permission notice shall be
-- included in all copies or substantial portions of the Software.
--
-- THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND,
-- EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF
-- MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND
-- NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS
-- BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN
-- ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN
-- CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
-- SOFTWARE.

local version = "0.9"
local copyright = "luap " .. version ..
   " Copyright (C) 2012-2023 Dimitris Papavasiliou, Boris Nagaev"

local prompt = require "prompt"
local argparse = require "argparse"

-- Define the command-line argument parser.

local parser = argparse()
   :name "luap"
   :description "A Lua command prompt with pretty-printing and auto-completion."
   :add_help "-h"
   :add_complete({hidden=true})

-- Options

parser:option "-e"
   :argname "STMT"
   :description "Execute string 'STMT'."
   :count "*"

parser:option "-l"
   :argname "NAME"
   :description "Require library 'NAME'."
   :count "*"

if jit then
   parser:option "-j"
      :argname "CMD"
      :description "Perform LuaJIT control command."
      :count "*"
   parser:option "-O"
      :argname "OPT"
      :description "Control LuaJIT optimizations."
      :count "*"
end

-- Flags

parser:flag "-p"
   :description "Force plain, uncolored output."

parser:flag "-v"
   :description "Print version information."

parser:flag "-i"
   :description "Enter interactive mode."

-- Arguments

parser:argument "SCRIPT"
   :description [[A Lua script to be executed.  Any arguments
specified after the script name, are passed to
the script.]]
   :args '*'

local args = parser:parse(arg)

-- Print version information and exit.

if args.v then
   print(table.concat ({copyright,
                        prompt.copyrights[1],
                        prompt.copyrights[2]}, "\n"))
   os.exit(0)
end

-- Pass optimization options to LuaJIT.

if args.O then
   for _, O in ipairs(args.O) do
      jit.opt.start(O)
   end
end

local interactive = (args.i or (prompt.interactive and
                                   #args.SCRIPT == 0 and #args.e == 0))

if interactive then
   print(table.concat ({prompt.copyrights[2], copyright}, "\n"))
end

-- Parse control commands and pass them to LuaJIT.

if args.j then
   for _, j in ipairs(args.j) do
      local unpack = unpack or table.unpack

      -- Parse the command name.

      local name = j:match('^[^=]+')

      if not name then
         print(parser:get_help())
         os.exit(0)
      end

      -- Parse the arguments, if any.

      local args = {}

      for arg in (j:match('=(.*)$') or ""):gmatch('[^,]+') do
         table.insert(args, arg)
      end

      -- Look for a builtin command.

      if jit[name] then
         jit[name](unpack(args))
      else
         local ok, m = pcall(require, 'jit.' .. name)

         if not ok or not m then
            parser:error('unknown luaJIT command or ' ..
                     'jit.* modules not installed')
         end

         m.start(unpack(args))
      end
   end
end

-- Require modules specified on the command line.

if #args.l > 0 then
   for _, l in ipairs(args.l) do
      if _VERSION == "Lua 5.1" then
         require(l)
      else
         _G[l] = require(l)
      end
   end
end

-- Load and execute chunks passed on the command line.

if #args.e > 0 then
   local loadstring = loadstring or load
   for _, e in ipairs(args.e) do
      loadstring(e)()
   end
end

-- Run the script given on the command line, passing any arguments as
-- required.

if #args.SCRIPT > 0 or (not interactive and #args.e == 0) then
   local chunk
   local loadstring = loadstring or load
   local unpack = unpack or table.unpack
   local name

   if #args.SCRIPT > 0 then
     name = table.remove(args.SCRIPT, 1)
   else
      name = "-"
   end

   if name == "-" then
      chunk, message = loadstring(io.stdin:read("*a"))
   else
      chunk, message = loadfile(name)
   end

   if chunk then
      -- This duplicates the behavior of the standard Lua interpreter
      -- to some extent.  Arguments prior to the script name are not
      -- passed.

      arg = {[0] = name, unpack(args.SCRIPT)}
      prompt.call(chunk, unpack(args.SCRIPT))
   else
      print(message)
      os.exit(0)
   end
end

-- Enter interactive mode, if appropriate.

if interactive then
   prompt.name = 'lua'
   prompt.prompts = {'>  ', '>> '}
   prompt.colorize = not args.p
   prompt.history = os.getenv('HOME') .. '/.lua_history'

   for _, name in ipairs{os.getenv('HOME') .. '/.luaprc.lua',
                         os.getenv('HOME') .. '/.config/luaprc.lua'} do

      local f = io.open(name, "r")
      if f ~= nil then
         io.close(f)

         chunk, message = loadfile(name)

         if chunk then
            chunk()
         else
            print(message)
         end

         break
      end
   end

   prompt.enter()
end
